gnucash/import-export/import-main-matcher.c

```diff
@@ -1435,38 +1499,151 @@ void gnc_gen_trans_list_add_trans_with_ref_id (GNCImportMainMatcher *gui, Transa
+{
+    split_find_match (s->transaction_info, data, s->display_threshold, s->fuzzy_amount);
+}
+
+/* This goes through the temporary list of imported transactions, runs a single query to gater potential matching
+ * register transactions based on accounts and dates, then selects potential matching transactions for
+ * each imported transaction. Doing a single query makes things a lot faster than doing one query per
+ * imported transaction.
+ */
+void
+gnc_gen_trans_list_create_matches (GNCImportMainMatcher *gui)
+{
+    GtkTreeModel* model = gtk_tree_view_get_model (gui->view);
+    GtkListStore* store = GTK_LIST_STORE (model);
+    GtkTreeIter iter;
+    GNCImportMatchInfo *selected_match;
+    gboolean match_selected_manually;
+    GHashTable* lists_per_accounts = NULL;
+    Query *query;
+    gint match_date_hardlimit = gnc_import_Settings_get_match_date_hardlimit (gui->user_settings);
+    gint display_threshold = gnc_import_Settings_get_display_threshold (gui->user_settings);
+    double fuzzy_amount = gnc_import_Settings_get_fuzzy_amount (gui->user_settings);
+    GSList* imported_trans;
+
+    query = qof_query_create_for (GNC_ID_SPLIT);
+    // Gather the list of splits that could match one of the imported transactions, this return a hash table.
+    lists_per_accounts = create_list_of_potential_matches (gui, model, query, match_date_hardlimit);
+
+    // For each imported transaction, grab the list of potential matches corresponding to that account,
+    // and evaluate how good a match they are.
+    for (imported_trans=gui->temp_trans_list; imported_trans!=NULL; imported_trans=imported_trans->next)
+    {
+        GNCImportTransInfo* transaction_info = imported_trans->data;
+        Account *importaccount = xaccSplitGetAccount (gnc_import_TransInfo_get_fsplit (transaction_info));
+        match_struct s = {transaction_info, display_threshold, fuzzy_amount};
+        
+        g_slist_foreach (g_hash_table_lookup (lists_per_accounts, importaccount), (GFunc) match_helper, &s);
+
+        // Sort the matches, select the best match, and set the action.
+        gnc_import_TransInfo_init_matches (transaction_info, gui->user_settings);
+
+        selected_match = gnc_import_TransInfo_get_selected_match (transaction_info);
+        match_selected_manually = gnc_import_TransInfo_get_match_selected_manually (transaction_info);
 
         if (selected_match)
             gnc_import_PendingMatches_add_match (gui->pending_matches,
                                                  selected_match,
                                                  match_selected_manually);
-
-        model = gtk_tree_view_get_model (gui->view);
-        gtk_tree_store_append (GTK_TREE_STORE(model), &iter, NULL);
+        
+        gtk_tree_store_append (GTK_TREE_STORE (model), &iter, NULL);
         refresh_model_row (gui, model, &iter, transaction_info);
     }
+    
+    qof_query_destroy (query);
+    g_slist_free (gui->temp_trans_list);
```

Comment:
This leaks the `trans_info`s. You need a `trans_info_delete` function and use `g_slist_free_full`.

Reply:
The freeing of trans_info is done in gnc_gen_trans_list_delete(), but your comment made me realize that the destruction of this list is better placed in gnc_gen_trans_list_delete().